modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala

```diff
@@ -74,10 +91,48 @@ object CurrencySnapshotAcceptanceManager {
         rewards
       )
 
+      lastMessages = lastSnapshotContext.snapshotInfo.lastMessages.getOrElse(SortedMap.empty[MessageType, Signed[CurrencyMessage]])
+
+      messagesToAccept = messagesForAcceptance
+        .groupBy(_.messageType)
+        .view
+        .mapValues { messages =>
+          val sortedMessages = messages.sortBy(_.parentOrdinal)
+
+          sortedMessages
+            .foldLeft(List.empty[Signed[CurrencyMessage]]) { (acc, curr) =>
+              lastMessages.get(curr.messageType) match {
+                case Some(lastMsg) if curr.parentOrdinal <= lastMsg.parentOrdinal => acc
+                case _ if acc.isEmpty || acc.last.ordinal == curr.parentOrdinal   => curr :: acc
+                case _                                                            => acc
+              }
+            }
+        }
+        .toMap
+        .collect {
+          case (k, x :: xs) => (k, x :: xs)
+        }
+
+      messagesToReject = messagesForAcceptance.filterNot(messagesToAccept.values.toList.flatten.contains)
+
+      messagesForContextUpdate = (lastMessages.toList ++ messagesToAccept.toList.flatMap {
+        case (t, xs) => xs.map(x => (t, x))
+      }).groupBy { case ((messageType, _)) => messageType }.view.mapValues {
+        _.map { case ((_, messages)) => messages }.maxByOption(_.ordinal)
+      }.toMap.collect {
+        case (k, Some(v)) => k -> v
+      }.toSortedMap
```

Comment:
The formatting makes this a bit hard to follow but is this refactor equivalent to what we're trying to do here?
```
messagesForContextUpdate =
  messagesToAccept
    .view
    .mapValues(_.maxByOption(_.ordinal))
    .foldLeft(lastMessages) {
      case (acc, (k, v)) =>
        acc.updatedWith(k) {
          case None => v
          case x@Some(v0) => v.filter(_.ordinal > v0.ordinal).orElse(x)
        }
    }
```

Reply:
well, I do the following steps:
1. group by MessageType
2. take a single message from the list by the max ordinal
3. filter out None (if there is no message for the message type

Reply:
Ok, so I think I understood the intent. In that case, isn't the above refactor a simplification? In effect,

1. For all messages to accept, find the max message for each message type.
2. Merge the result with `lastMessages`

The result from step 1 has the same type as `lastMessages` so I think we can avoid most of the intermediary data structures.

Reply:
Working on immutable data structures, we create new instances anyway. Does it really matter it's this or that type as an intermediate type? I think we need to select between readability and efficiency and in this case we simply follow the easy algorithm and the second case requires an additional logic for merging both structures (instead of merging first and then operating on all messages) 🤔

Reply:
Ok. I see I can use a semigroup to make it easier and more readable:
1. merge 
2. take highest by ordinal
3. filter-out None

![image](https://github.com/Constellation-Labs/tessellation/assets/4959931/2602c2f2-cca9-4bdf-b2a8-4a105dc8d528)